indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java

```diff
@@ -749,131 +651,69 @@ private void handleStatus(final TaskStatus status)
+   * even if the latter is manually modified.
+   * <p>
+   * This method must be called only when queue is {@link #active}, except when
+   * starting up.
    */
-  private void syncFromStorage()
+  private synchronized void syncFromStorage()
   {
-    giant.lock();
-
-    try {
-      if (active) {
-        final Map<String, Task> newTasks = toTaskIDMap(taskStorage.getActiveTasks());
-        final int tasksSynced = newTasks.size();
-        final Map<String, Task> oldTasks = new HashMap<>(tasks);
-
-        // Calculate differences on IDs instead of Task Objects.
-        Set<String> commonIds = Sets.newHashSet(Sets.intersection(newTasks.keySet(), oldTasks.keySet()));
-        for (String taskID : commonIds) {
-          newTasks.remove(taskID);
-          oldTasks.remove(taskID);
-        }
-        Collection<Task> addedTasks = newTasks.values();
-        Collection<Task> removedTasks = oldTasks.values();
-
-        // Clean up removed Tasks
-        for (Task task : removedTasks) {
-          removeTaskInternal(task.getId());
-        }
-
-        // Add newly Added tasks to the queue
-        for (Task task : addedTasks) {
-          addTaskInternal(task);
-        }
-
-        log.info(
-            "Synced %d tasks from storage (%d tasks added, %d tasks removed).",
-            tasksSynced,
-            addedTasks.size(),
-            removedTasks.size()
-        );
-        requestManagement();
-      } else {
-        log.info("Not active. Skipping storage sync.");
-      }
-    }
-    catch (Exception e) {
-      log.warn(e, "Failed to sync tasks from storage!");
-      throw new RuntimeException(e);
-    }
-    finally {
-      giant.unlock();
-    }
-  }
-
-  private static Map<String, Task> toTaskIDMap(List<Task> taskList)
-  {
-    Map<String, Task> rv = new HashMap<>();
-    for (Task task : taskList) {
-      rv.put(task.getId(), task);
-    }
-    return rv;
-  }
-
-  private Map<String, Long> getDeltaValues(Map<String, Long> total, Map<String, Long> prev)
-  {
-    return total.entrySet()
-                .stream()
-                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue() - prev.getOrDefault(e.getKey(), 0L)));
-  }
+    final Map<String, Task> newTasks = taskStorage.getActiveTasks().stream().collect(
+        Collectors.toMap(Task::getId, Function.identity())
+    );
+    final Map<String, Task> oldTasks = new HashMap<>(tasks);
+
+    // Calculate differences on IDs instead of Task Objects.
+    Set<String> commonIds = Sets.intersection(newTasks.keySet(), oldTasks.keySet());
+    for (String taskId : commonIds) {
+      newTasks.remove(taskId);
+      oldTasks.remove(taskId);
+    }
+    Collection<Task> addedTasks = newTasks.values();
+    Collection<Task> removedTasks = oldTasks.values();
+
+    // Add new tasks and clean up removed tasks
+    addedTasks.forEach(this::addTaskInternal);
+    removedTasks.forEach(this::removeTaskInternal);
+    log.info(
```

Comment:
This existed before, but it might be wise to remove as much as possible, including logging, from the `synchronized` block, and tighten it down a bit.

Reply:
I don't think the footprint of this call would really make much of a difference compared to the earlier operations, fetching from DB and map manipulation.

But yeah, you never really know with logging. I suppose I could just return the counts from this method and log the values in the caller itself.

Reply:
Could use `synchronized (this) { .... }` ?

Reply:
But yes - it does raise a good qn about the remainder of the workflow.